README.md
# CPL
Calmira PLatform - минимальный набор скомпилированного программного обеспечения для сборки дистрибутива CalmiraLinux. Это *не* загрузочные образы системы.

![Лого](design/calmira_logo_transparent_black.png)

## Структура
Монолитная. Представляет единый архив с нужными ПО и файлами.

Программное обеспечение, входящее в состав платформы, не может быть удалено стандартными средствами Calmira, но может быть подменено или модифицировано. В любом случае, остаётся возможность ручного удаления ненужных файлов.

Пакеты, а так же информация о каждом доступна в системе портов.

## Обновления программного обеспечения
Сборки CPL формируются к выходу крупной версии дистрибутива [CalmiraLinux](https://github.com/CalmiraLinux/CalmiraLinux). В рамках Calmira может быть обновлён или изменён *любой* пакет из платформы.

## Список ПО
### Временная система
> Ожидайте в скором времени

### Платформа
> Ожидайте в скором времени


## Предоставляемый вспомогательный инструментарий

* `makesystem2` - скрипт для автоматизированной сборки системы. Входит в состав CABS.

CHANGELOG.md
# ChangeLog

* 09.11.2021 Михаил Краснов <linuxoid85@gmail.com>
	* **Обновления пакетов:**
		* automake-1.16.5;
		* bc-5.1.1;
		* bison-3.8.2;
		* coreutils-9.0;
		* diffutils-3.8;
		* e2fsprogs-1.46.4;
		* file-5.41;
		* gawk-5.1.1;
		* gcc-11.2.0;
		* gdbm-1.22;
		* glibc-2.34;
		* grep-3.7;
		* gzip-1.11;
		* iana-etc-20211012;
		* inetutils-2.2;
		* iproute2-5.15.0;
		* libpipeline-1.5.4;
		* libcap-2.60;
		* linux-5.15.1;
		* meson-0.60.1;
		* ncurses-6.3;
		* openssl-3.0.0;
		* python-3.10;
		* shadow-4.9;
		* tzdata-2021d;
		* zlib-ng-2.0.5.

* 07.11.2021 Михаил Краснов <linuxoid85@gmail.com>
	* Фикс сборки texinfo;

* 20.10.2021 Михаил Краснов <linuxoid85@gmail.com>
	* Добавление поддержки CPL.

* 18.10.2021 Михаил Краснов <linuxoid85@gmail.com>
	* Фикс ошибок в коде `makesystem2`; добавление перенаправления в лог файл.

* 16.10.2021 Михаил Краснов <linuxoid85@gmail.com>
	* Добавление инструментария `mksys2` для автоматизированной сборки системы.
	* Initial commit.


src/makesys/src/makesys.py
#!/usr/bin/python3
# Скрипт для автоматизации сборки системы
# (C) 2021 Михаил Краснов <linuxoid85@gmail.com>
# Для Calmira LX4 1.1 GNU/Linux
# Версия: print(VERSION)

## TODOS ##
# TODO - Add build time analysis and redirect messages to the log for all build scripts

## Imports ##
import os
import sys
import json
import subprocess

## Base constants ##
VERSION = "0.2"
JSON_FILE = "/etc/makesys.json"
BuildInstructions = "/usr/src/packages/" # Files with build instructions
LogDir = "/var/log/system_building/packages"
LogFile = "/var/log/system_building/system_building.log"

default_message = "Continue? (y/n) "

## Base functions ##
# Заголовок
def header_msg(message):
    print("\033[1m{}\033[0m".format(message))

# Отправка сообщений в лог
def log_msg(message):   
    f = open(LogFile, 'a')
    message = message + "\n"
    for index in message:
        f.write(index)
    f.write('\n')
    f.close()

# Диалог с пользователем
def dialog_msg(message=default_message, return_code=0):
    print(message, end=" ")
    run = input()
    
    if run == "y" or run == "Y":
        print("Продолжается сборка...")
    elif run == "n" or run == "N":
        print("Сборка прервана!")
        exit(return_code)
    else:
        print("НЕПРАВИЛЬНОЕ ЗНАЧЕНИЕ ({}). Аварийное завершение работы.".format(run))


## Script body ##

# Check root
GID = os.getgid()
if GID != 0:
    print("\033[31mERROR: you must run this script on root! \033[0m")
    exit(1)

# Check sys.argv[1]
if sys.argv[1] != "sysvinit" or sys.argv[1] != "cpl":
    print("Uknown build mode {}!".format(sys.argv[1]))
    exit(1)

if sys.argv[1] == "cpl":
    cpl_build = True

# Check log dir
if os.path.isdir(LogDir):
    pass
else:
    print("Create directory {}...".format(LogDir))
    os.makedirs(LogDir)

header_msg("Start base system building")
log_msg("START SYSTEM BUILDING")

if os.path.isfile(JSON_FILE):
    pass
else:
    print("Error: file {} not found".format(JSON_FILE))
    exit(1)

with open(JSON_FILE) as f:
    fileData = json.load(f)

# Проверка на существование файлов
errorPkgs = 0
for File in fileData["files"]:
    PackageFile = BuildInstructions + File

    print("Testing file {}...".format(PackageFile), end = " ")
    if os.path.isfile(PackageFile):
        print("\033[32mok\033[0m")
    else:
        print("\033[31mFAIL\033[0m")
        errorPkgs += 1

if errorPkgs > 0:
    print("No {} build scripts found!".format(errorPkgs))
    sys.exit(1)
    
"""
Building a system.

The required script with assembly instructions is launched.

If it returned 0, then the build was successful. If it returned a value
other than zero, then a warning about incorrect building is displayed.

Usage:

`build_package(mode)` - for building base system;
`build_package(init="init")` - for building system init.

Inits:
- sysvinit;
- systemd.
"""
def build_package(mode, init="sysvinit"):
    if mode == "system":
        build_files = fileData["packages"]
        
    elif mode == "init":
        if init == "sysvinit":
            build_files = fileData["systemv-packages"]
            
        else:
            print("Uknown init system {} for building!".format(init))
            exit(1)
    
    elif mode == "cpl":
        build_files = fileData["cpl-system"]

    else:
        print("Uknown mode for 'build_package()'!")
        exit(1)
            
    for file in build_files:
        print("\033[35mBuilding package\033[0m {}...".format(file))
        
        PackageFile = BuildInstructions + file + "2>&1 | tee " + LogDir + "/" + file
        log_message = "Build package " + file
        
        log_msg(log_message)
        
        result = subprocess.run(PackageFile, shell = True)
        if result.returncode == 0:
            print("\033[1mPackage {} returned 0, which means the build is probally seccessful.".format(file))
        else:
            print("\033[1mPackage {0} returned {1}, which means the assembly is probably WRONG!".format(file, result.returncode))
            dialog_msg(return_code=1)

# Building base system
if cpl_build == True:
    build_package("cpl")
else:
    build_package("system")

# Building system init
build_package("init", init = sys.argv[1])

print("************************************")
print("**** Base system build complete ****")
print("************************************")


src/makesystem2/src/new/mksys.py
#!/usr/bin/python3

import os
import sys
import subprocess
import json
import argparse

## BASE CONSTANTS ##
SETTINGS_DIR       = "./settings" # Settings directory
SETTINGS_FILE      = SETTINGS_DIR + "/makesystem.json"    # Program settings
CPL_FILE           = SETTINGS_DIR + "/make-cpl.json"      # Building the CPL
POST_CPL_FILE      = SETTINGS_DIR + "/make-post-cpl.json" # Building the post-CPL
POST_CPL_SYSV_FILE = SETTINGS_DIR + "/make-sysvinit.json" # Building the post-CPL with sysvinit
POST_CPL_SYSD_FILE = SETTINGS_DIR + "/make-systemd.json"  # Building the port-CPL with systemd
BUILD_INSTRUCTIONS = "./packages"                         # Directory with packages build instructions
LOG_FILE           = "/var/log/system_building.log"       # Log file
DEFAULT_MESSAGE    = "Continue? (y/n)"                    # For dialog_msg()

## BASE FUNCTIONS ##
def log_msg(message):
    f = open(LOG_FILE, "a")
    message = message + "\n"
    for index in message:
        f.write(index)
    f.close()

# Dialog
def dialog_msg(message=DEFAULT_MESSAGE, return_code=0):
    print(message, end=" ")
    run = input()

    if run == "y" or run == "Y":
        print("Continue building...")
    elif run == "n" or run == "N":
        print("Aborted!")
        exit(return_code)
    else:
        print("Uknown input! Aborted.")
        exit(1)

# System building
def build_packages(mode="post-cpl", init="sysvinit"):
    if mode == "post-cpl":
        match init:
            case "sysvinit":
                f = open(POST_CPL_SYSV_FILE, "r")
                packageData = json.load(f)
            case "systemd":
                f = open(POST_CPL_SYSD_FILE, "r")
                packageData = json.load(f)
            default:
                print("Init choosing error!")
                exit(1)
    elif mode == "cpl":
        f = open(CPL_FILE, "r")
        packageData = json.load(f)
    else:
        print("Building mode choosing error!")
        exit(1)
    
    for package in packageData["package-sequence"]:
        files = packageData["packages"][package]["build_sequence"]
        
        for build_file in files:
            run = subprocess.run(build_file, shell=True)
            
            if run == 0:
                print("Package {} build OK".format(package))
            else:
                print("Package {} build FAIL".format(package))


src/makesystem2/src/makesystem.py
#!/usr/bin/python3
# Скрипт для автоматизации сборки системы
# (C) 2021 Михаил Краснов <linuxoid85@gmail.com>
# Для Calmira LX4 1.1 GNU/Linux
# Версия: print(VERSION)

## TODOS ##
# TODO - Add build time analysis and redirect messages to the log for all build scripts

## Imports ##
import os
import sys
import json
import subprocess

## Base constants ##
VERSION = "0.2"
JSON_FILE = "/etc/makesystem.json"
BuildInstructions = "/usr/src/packages/" # Files with build instructions
LogDir = "/var/log/system_building/packages"
LogFile = "/var/log/system_building/system_building.log"

default_message = "Continue? (y/n) "

## Base functions ##
# Заголовок
def header_msg(message):
    print("\033[1m{}\033[0m".format(message))

# Отправка сообщений в лог
def log_msg(message):   
    f = open(LogFile, 'a')
    message = message + "\n"
    for index in message:
        f.write(index)
    f.write('\n')
    f.close()

# Диалог с пользователем
def dialog_msg(message=default_message, return_code=0):
    print(message, end="")
    run = input()
    
    if run == "y" or run == "Y":
        print("Продолжается сборка...")
    elif run == "n" or run == "N":
        print("Сборка прервана!")
        exit(return_code)
    else:
        print("НЕПРАВИЛЬНОЕ ЗНАЧЕНИЕ ({}). Аварийное завершение работы.".format(run))


## Script body ##

# Check root
GID = os.getgid()
if GID != 0:
    print("\033[31mERROR: you must run this script on root! \033[0m")
    exit(1)

# Check sys.argv[1]
if sys.argv[1] != "systemd" or sys.argv[1] != "sysvinit" or sys.argv[1] != "cpl":
    print("Uknown build mode {}!".format(sys.argv[1]))
    exit(1)

if sys.argv[1] == "cpl":
    cpl_build = True

# Check log dir
if os.path.isdir(LogDir):
    pass
else:
    print("Create directory {}...".format(LogDir))
    os.makedirs(LogDir)

header_msg("Start base system building")
log_msg("START SYSTEM BUILDING")

if os.path.isfile(JSON_FILE):
    pass
else:
    print("Error: file {} not found".format(JSON_FILE))
    exit(1)

with open(JSON_FILE) as f:
    fileData = json.load(f)

# Проверка на существование файлов
errorPkgs = 0
for File in fileData["files"]:
    PackageFile = BuildInstructions + File

    print("Testing file {}...".format(PackageFile), end = " ")
    if os.path.isfile(PackageFile):
        print("\033[32mok\033[0m")
    else:
        print("\033[31mFAIL\033[0m")
        errorPkgs += 1

if errorPkgs > 0:
    print("No {} build scripts found!".format(errorPkgs))
    sys.exit(1)
    
"""
Building a system.

The required script with assembly instructions is launched.

If it returned 0, then the build was successful. If it returned a value
other than zero, then a warning about incorrect building is displayed.

Usage:

`build_package(mode)` - for building base system;
`build_package(init="init")` - for building system init.

Inits:
- sysvinit;
- systemd.
"""
def build_package(mode, init="sysvinit"):
    if mode == "system":
        build_files = fileData["packages"]
        
    elif mode == "init":
        if init == "sysvinit":
            build_files = fileData["systemv-packages"]
            
        elif init == "systemd":
            build_files = fileData["systemd-packages"]
            
        else:
            print("Uknown init system {} for building!".format(init))
            exit(1)
    
    elif mode == "cpl":
        build_files = fileData["cpl-system"]

    else:
        print("Uknown mode for 'build_package()'!")
        exit(1)
            
    for file in build_files:
        print("\033[35mBuilding package\033[0m {}...".format(file))
        
        PackageFile = BuildInstructions + file + "2>&1 | tee " + LogDir + "/" + file
        log_message = "Build package " + file
        
        log_msg(log_message)
        
        result = subprocess.run(PackageFile, shell = True)
        if result.returncode == 0:
            print("\033[1mPackage {} returned 0, which means the build is probally seccessful.".format(file))
        else:
            print("\033[1mPackage {0} returned {1}, which means the assembly is probably WRONG!".format(file, result.returncode))
            dialog_msg(return_code=1)

# Building base system
if cpl_build == True:
    build_package("cpl")
else:
    build_package("system")

# Building system init
build_package("init", init = sys.argv[1])

print("************************************")
print("**** Base system build complete ****")
print("************************************")